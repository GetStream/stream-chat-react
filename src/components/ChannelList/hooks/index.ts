export * from './useChannelDeletedListener';
export * from './useChannelHiddenListener';
export * from './useChannelTruncatedListener';
export * from './useChannelUpdatedListener';
export * from './useChannelVisibleListener';
export * from './useConnectionRecoveredListener';
export * from './useMessageNewListener';
export * from './useMobileNavigation';
export * from './useNotificationAddedToChannelListener';
export * from './useNotificationMessageNewListener';
export * from './useNotificationRemovedFromChannelListener';
export * from './usePaginatedChannels';
export * from './useUserPresenceChangedListener';
export * from './useChannelMembershipState';
