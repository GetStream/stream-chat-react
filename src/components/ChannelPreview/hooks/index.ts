export { useChannelPreviewInfo } from './useChannelPreviewInfo';
export { MessageDeliveryStatus } from './useMessageDeliveryStatus';
