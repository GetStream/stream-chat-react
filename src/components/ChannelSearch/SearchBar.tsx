import React, {
  MouseEventHandler,
  PropsWithChildren,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import clsx from 'clsx';

import {
  MenuIcon as DefaultMenuIcon,
  SearchIcon as DefaultSearchInputIcon,
  ReturnIcon,
  XIcon,
} from './icons';
import { SearchInput as DefaultSearchInput, SearchInputProps } from './SearchInput';

export type AppMenuProps = {
  close?: () => void;
};

type SearchBarButtonProps = {
  className?: string;
  onClick?: MouseEventHandler<HTMLButtonElement>;
};

const SearchBarButton = ({
  children,
  className,
  onClick,
}: PropsWithChildren<SearchBarButtonProps>) => (
  <button
    className={clsx('str-chat__channel-search-bar-button', className)}
    data-testid='search-bar-button'
    onClick={onClick}
  >
    {children}
  </button>
);

export type SearchBarController = {
  /** Called on search input focus */
  activateSearch: () => void;
  /** Clears the search state, removes focus from the search input */
  exitSearch: () => void;
  /** Flag determining whether the search input is focused */
  inputIsFocused: boolean;
  /** Ref object for the input wrapper in the SearchBar */
  searchBarRef: React.RefObject<HTMLDivElement>;
};

export type AdditionalSearchBarProps = {
  /** Application menu to be displayed  when clicked on MenuIcon */
  AppMenu?: React.ComponentType<AppMenuProps>;
  /** Custom icon component used to clear the input value on click. Displayed within the search input wrapper. */
  ClearInputIcon?: React.ComponentType;
  /** Custom icon component used to terminate the search UI session on click. */
  ExitSearchIcon?: React.ComponentType;
  /** Custom icon component used to invoke context menu. */
  MenuIcon?: React.ComponentType;
  /** Custom UI component to display the search text input */
  SearchInput?: React.ComponentType<SearchInputProps>;
  /** Custom icon used to indicate search input. */
  SearchInputIcon?: React.ComponentType;
};

export type SearchBarProps = AdditionalSearchBarProps & SearchBarController & SearchInputProps;

// todo: add context menu control logic
export const SearchBar = (props: SearchBarProps) => {
  const {
    activateSearch,
    AppMenu,
    ClearInputIcon = XIcon,
    exitSearch,
    ExitSearchIcon = ReturnIcon,
    inputIsFocused,
    MenuIcon = DefaultMenuIcon,
    searchBarRef,
    SearchInput = DefaultSearchInput,
    SearchInputIcon = DefaultSearchInputIcon,
    ...inputProps
  } = props;

  const [menuIsOpen, setMenuIsOpen] = useState(false);
  const appMenuRef = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    if (!appMenuRef.current) return;

    const handleKeyDown = (event: KeyboardEvent) => {
      if (menuIsOpen && event.key === 'Escape') {
        setMenuIsOpen(false);
      }
    };

    const clickListener = (e: MouseEvent) => {
      if (
        !(e.target instanceof HTMLElement) ||
        !menuIsOpen ||
        appMenuRef.current?.contains(e.target)
      )
        return;
      setMenuIsOpen(false);
    };

    document.addEventListener('keydown', handleKeyDown);
    document.addEventListener('click', clickListener);
    return () => {
      document.removeEventListener('keydown', handleKeyDown);
      document.removeEventListener('click', clickListener);
    };
  }, [menuIsOpen]);

  useEffect(() => {
    if (!props.inputRef.current) return;

    const handleFocus = () => {
      activateSearch();
    };

    const handleBlur = (e: Event) => {
      e.stopPropagation(); // handle blur/focus state with React state
    };

    props.inputRef.current.addEventListener('focus', handleFocus);
    props.inputRef.current.addEventListener('blur', handleBlur);
    return () => {
      props.inputRef.current?.removeEventListener('focus', handleFocus);
      props.inputRef.current?.addEventListener('blur', handleBlur);
    };
  }, []);

  const handleClearClick = useCallback(() => {
    exitSearch();
    inputProps.inputRef.current?.focus();
  }, []);

  const closeAppMenu = useCallback(() => setMenuIsOpen(false), []);

  return (
    <div className='str-chat__channel-search-bar' data-testid='search-bar' ref={searchBarRef}>
      {inputIsFocused ? (
        <SearchBarButton
          className='str-chat__channel-search-bar-button--exit-search'
          onClick={exitSearch}
        >
          <ExitSearchIcon />
        </SearchBarButton>
      ) : AppMenu ? (
        <SearchBarButton
          className='str-chat__channel-search-bar-button--menu'
          onClick={() => setMenuIsOpen((prev) => !prev)}
        >
          <MenuIcon />
        </SearchBarButton>
      ) : null}

      <div
        className={clsx(
          'str-chat__channel-search-input--wrapper',
          inputProps.query && 'str-chat__channel-search-input--wrapper-active',
        )}
      >
        <div className='str-chat__channel-search-input--icon'>
          <SearchInputIcon />
        </div>
        <SearchInput {...inputProps} />
        <button
          className='str-chat__channel-search-input--clear-button'
          data-testid='clear-input-button'
          disabled={!inputProps.query}
          onClick={handleClearClick}
        >
          <ClearInputIcon />
        </button>
      </div>
      {menuIsOpen && AppMenu && (
        <div ref={appMenuRef}>
          <AppMenu close={closeAppMenu} />
        </div>
      )}
    </div>
  );
};
