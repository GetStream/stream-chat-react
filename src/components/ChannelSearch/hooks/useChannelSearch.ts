import React, { useCallback, useEffect, useRef, useState } from 'react';
import throttle from 'lodash.throttle';
import uniqBy from 'lodash.uniqby';

import { ChannelOrUserResponse, isChannel } from '../utils';

import { useChatContext } from '../../../context/ChatContext';

import type {
  ChannelFilters,
  ChannelOptions,
  ChannelSort,
  UserFilters,
  UserOptions,
  UserSort,
} from 'stream-chat';

import type { Channel } from 'stream-chat';
import type { SearchBarController } from '../SearchBar';
import type { SearchInputController } from '../SearchInput';
import type { SearchResultsController } from '../SearchResults';
import type { DefaultStreamChatGenerics } from '../../../types/types';

export type ChannelSearchFunctionParams<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  setQuery: React.Dispatch<React.SetStateAction<string>>;
  setResults: React.Dispatch<React.SetStateAction<ChannelOrUserResponse<StreamChatGenerics>[]>>;
  setSearching: React.Dispatch<React.SetStateAction<boolean>>;
};

export type SearchController<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = SearchInputController & SearchBarController & SearchResultsController<StreamChatGenerics>;

export type SearchQueryParams<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  channelFilters?: {
    filters?: ChannelFilters<StreamChatGenerics>;
    options?: ChannelOptions;
    sort?: ChannelSort<StreamChatGenerics>;
  };
  userFilters?: {
    filters?:
      | UserFilters<StreamChatGenerics>
      | ((query: string) => UserFilters<StreamChatGenerics>);
    options?: UserOptions;
    sort?: UserSort<StreamChatGenerics>;
  };
};

export type ChannelSearchParams<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  /** The type of channel to create on user result select, defaults to `messaging` */
  channelType?: string;
  /** Clear search state / results on every click outside the search input, defaults to true */
  clearSearchOnClickOutside?: boolean;
  /** Disables execution of the search queries, defaults to false */
  disabled?: boolean;
  /** Callback invoked with every search input change handler */
  onSearch?: SearchInputController['onSearch'];
  /** Callback invoked when the search UI is deactivated */
  onSearchExit?: () => void;
  /** Custom handler function to run on search result item selection */
  onSelectResult?: (
    params: ChannelSearchFunctionParams<StreamChatGenerics>,
    result: ChannelOrUserResponse<StreamChatGenerics>,
  ) => Promise<void> | void;
  /** Boolean to search for channels as well as users in the server query, default is false and just searches for users */
  searchForChannels?: boolean;
  /** Custom search function to override the default implementation */
  searchFunction?: (
    params: ChannelSearchFunctionParams<StreamChatGenerics>,
    event: React.BaseSyntheticEvent,
  ) => Promise<void> | void;
  /** Object containing filters/sort/options overrides for user / channel search */
  searchQueryParams?: SearchQueryParams<StreamChatGenerics>;
};

export type ChannelSearchControllerParams<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = ChannelSearchParams<StreamChatGenerics> & {
  /** Set the array of channels displayed in the ChannelList */
  setChannels: React.Dispatch<React.SetStateAction<Array<Channel<StreamChatGenerics>>>>;
};

export const useChannelSearch = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>({
  channelType = 'messaging',
  clearSearchOnClickOutside = true,
  disabled = false,
  onSearch: onSearchCallback,
  onSearchExit,
  onSelectResult,
  searchForChannels = false,
  searchFunction,
  searchQueryParams,
  setChannels,
}: ChannelSearchControllerParams<StreamChatGenerics>): SearchController<StreamChatGenerics> => {
  const { client, navOpen, setActiveChannel, themeVersion } = useChatContext<StreamChatGenerics>(
    'useChannelSearch',
  );

  const [inputIsFocused, setInputIsFocused] = useState(false);
  const [query, setQuery] = useState('');
  const [results, setResults] = useState<Array<ChannelOrUserResponse<StreamChatGenerics>>>([]);
  const [searching, setSearching] = useState(false);

  const inputRef = useRef<HTMLInputElement>(null);
  const searchBarRef = useRef<HTMLDivElement>(null);

  const clearState = useCallback(() => {
    setQuery('');
    setResults([]);
    setSearching(false);
  }, []);

  const activateSearch = useCallback(() => {
    setInputIsFocused(true);
  }, []);

  const exitSearch = useCallback(() => {
    setInputIsFocused(false);
    inputRef.current?.blur();
    clearState();
    onSearchExit?.();
  }, [clearState, onSearchExit]);

  useEffect(() => {
    if (disabled) return;

    const clickListener = (event: MouseEvent) => {
      if (!(event.target instanceof HTMLElement)) return;
      const isInputClick =
        themeVersion === '2'
          ? searchBarRef.current?.contains(event.target)
          : inputRef.current?.contains(event.target);

      if (isInputClick) return;

      if ((inputIsFocused && (!query || navOpen)) || clearSearchOnClickOutside) {
        exitSearch();
      }
    };

    document.addEventListener('click', clickListener);
    return () => document.removeEventListener('click', clickListener);
  }, [disabled, inputIsFocused, query, exitSearch, navOpen, clearSearchOnClickOutside]);

  useEffect(() => {
    if (!inputRef.current || disabled) return;

    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === 'Escape') return exitSearch();
    };
    inputRef.current.addEventListener('keydown', handleKeyDown);

    return () => {
      inputRef.current?.removeEventListener('keydown', handleKeyDown);
    };
  }, [disabled]);

  const selectResult = useCallback(
    async (result: ChannelOrUserResponse<StreamChatGenerics>) => {
      if (!client.userID) return;
      if (onSelectResult) {
        await onSelectResult(
          {
            setQuery,
            setResults,
            setSearching,
          },
          result,
        );
        return;
      }
      let selectedChannel: Channel<StreamChatGenerics>;
      if (isChannel(result)) {
        setActiveChannel(result);
        selectedChannel = result;
      } else {
        const newChannel = client.channel(channelType, { members: [client.userID, result.id] });
        await newChannel.watch();

        setActiveChannel(newChannel);
        selectedChannel = newChannel;
      }
      setChannels((channels) => uniqBy([selectedChannel, ...channels], 'cid'));
      if (clearSearchOnClickOutside) {
        exitSearch();
      }
    },
    [clearSearchOnClickOutside, client, exitSearch, onSelectResult, setActiveChannel, setChannels],
  );

  const getChannels = useCallback(
    async (text: string) => {
      if (!text || searching) return;
      setSearching(true);

      try {
        const userResponse = await client.queryUsers(
          // @ts-expect-error
          {
            $or: [{ id: { $autocomplete: text } }, { name: { $autocomplete: text } }],
            id: { $ne: client.userID },
            ...searchQueryParams?.userFilters?.filters,
          },
          { id: 1, ...searchQueryParams?.userFilters?.sort },
          { limit: 8, ...searchQueryParams?.userFilters?.options },
        );

        if (searchForChannels) {
          const channelResponse = client.queryChannels(
            // @ts-expect-error
            {
              name: { $autocomplete: text },
              ...searchQueryParams?.channelFilters?.filters,
            },
            searchQueryParams?.channelFilters?.sort || {},
            { limit: 5, ...searchQueryParams?.channelFilters?.options },
          );

          const [channels, { users }] = await Promise.all([channelResponse, userResponse]);

          setResults([...channels, ...users]);
          setSearching(false);
          return;
        }

        const { users } = await Promise.resolve(userResponse);

        setResults(users);
      } catch (error) {
        clearState();
        console.error(error);
      }

      setSearching(false);
    },
    [client, searching, searchForChannels],
  );

  const getChannelsThrottled = throttle(getChannels, 200);

  const onSearch = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      event.preventDefault();
      if (disabled) return;

      if (searchFunction) {
        searchFunction(
          {
            setQuery,
            setResults,
            setSearching,
          },
          event,
        );
      } else {
        setQuery(event.target.value);
        getChannelsThrottled(event.target.value);
      }
      onSearchCallback?.(event);
    },
    [disabled, getChannelsThrottled, onSearchCallback, searchFunction],
  );

  return {
    activateSearch,
    clearState,
    exitSearch,
    inputIsFocused,
    inputRef,
    onSearch,
    query,
    results,
    searchBarRef,
    searching,
    selectResult,
  };
};
