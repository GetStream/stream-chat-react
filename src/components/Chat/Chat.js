// @ts-check

import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';

import Dayjs from 'dayjs';
import { ChatContext, TranslationContext } from '../../context';
import { Streami18n } from '../../i18n';

import { version } from '../../../package.json';

/**
 * Chat - Wrapper component for Chat. The needs to be placed around any other chat components.
 * This Chat component provides the ChatContext to all other components.
 *
 * The ChatContext provides the following props:
 *
 * - client (the client connection)
 * - channels (the list of channels)
 * - setActiveChannel (a function to set the currently active channel)
 * - channel (the currently active channel)
 *
 * It also exposes the withChatContext HOC which you can use to consume the ChatContext
 *
 * @example ../../docs/Chat.md
 * @typedef {import('stream-chat').Channel | undefined} ChannelState
 * @type {React.FC<{ client: import('types').StreamChatReactClient, theme?: string, i18nInstance?: Streami18n, initialNavOpen?: boolean }>}
 */
const Chat = ({
  client,
  theme = 'messaging light',
  i18nInstance,
  initialNavOpen = true,
  children,
}) => {
  const [translators, setTranslators] = useState(
    /** @type { Required<import('types').TranslationContextValue>} */ ({
      t: /** @param {string} key */ (key) => key,
      tDateTimeParser: (input) => Dayjs(input),
    }),
  );
  const [mutes, setMutes] = useState(
    /** @type {import('stream-chat').Mute[]} */ ([]),
  );
  const [navOpen, setNavOpen] = useState(initialNavOpen);
  const [channel, setChannel] = useState(
    /** @type {ChannelState} */ (undefined),
  );

  const openMobileNav = () => setTimeout(() => setNavOpen(true), 100);
  const closeMobileNav = () => setNavOpen(false);
  const clientMutes = client?.user?.mutes;

  useEffect(() => {
    if (!client) return;
    const userAgent = client.getUserAgent();
    if (!userAgent.includes('stream-chat-react')) {
      // should result in something like:
      // 'stream-chat-react-2.3.2-stream-chat-javascript-client-browser-2.2.2'
      client.setUserAgent(`stream-chat-react-${version}-${userAgent}`);
    }
    // don't want client in dep array because it is a required
    // prop for this component and we only want this run on mount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [client]);

  useEffect(() => {
    setMutes(clientMutes || []);

    /** @param {import('stream-chat').Event} e */
    const handleEvent = (e) => {
      if (e.type === 'notification.mutes_updated') setMutes(e.me?.mutes || []);
    };
    if (client) client.on(handleEvent);
    return () => client && client.off(handleEvent);
  }, [client, clientMutes]);

  useEffect(() => {
    let streami18n;
    if (i18nInstance instanceof Streami18n) streami18n = i18nInstance;
    else streami18n = new Streami18n({ language: 'en' });

    streami18n.registerSetLanguageCallback((t) =>
      setTranslators((prevTranslator) => ({ ...prevTranslator, t })),
    );
    streami18n.getTranslators().then((translator) => {
      if (translator) setTranslators(translator);
    });
  }, [i18nInstance]);

  const setActiveChannel = useCallback(
    /**
     * @param {ChannelState} activeChannel
     * @param {{ limit?: number; offset?: number }} [watchers]
     * @param {React.BaseSyntheticEvent} [e]
     */
    async (activeChannel, watchers = {}, e) => {
      if (e && e.preventDefault) e.preventDefault();

      if (activeChannel && Object.keys(watchers).length) {
        await activeChannel.query({ watch: true, watchers });
      }
      setChannel(activeChannel);
      closeMobileNav();
    },
    [],
  );

  if (!translators.t) return null;

  return (
    <ChatContext.Provider
      value={{
        client,
        theme,
        channel,
        mutes,
        navOpen,
        setActiveChannel,
        openMobileNav,
        closeMobileNav,
      }}
    >
      <TranslationContext.Provider value={translators}>
        {children}
      </TranslationContext.Provider>
    </ChatContext.Provider>
  );
};

Chat.propTypes = {
  /** The StreamChat client object */
  client: /** @type {PropTypes.Validator<import('stream-chat').StreamChat>} */ (PropTypes
    .object.isRequired),
  /**
   *
   * Theme could be used for custom styling of the components.
   *
   * You can override the classes used in our components under parent theme class.
   *
   * e.g. If you want to build a theme where background of message is black
   *
   * ```
   *  <Chat client={client} theme={demo}>
   *    <Channel>
   *      <MessageList />
   *    </Channel>
   *  </Chat>
   * ```
   *
   * ```scss
   *  .demo.str-chat {
   *    .str-chat__message-simple {
   *      &-text-inner {
   *        background-color: black;
   *      }
   *    }
   *  }
   * ```
   *
   * Built in available themes:
   *
   *  - `messaging light`
   *  - `messaging dark`
   *  - `team light`
   *  - `team dark`
   *  - `commerce light`
   *  - `commerce dark`
   *  - `livestream light`
   *  - `livestream dark`
   */
  theme: PropTypes.string,
  /** navOpen initial status */
  initialNavOpen: PropTypes.bool,
};

export default Chat;
