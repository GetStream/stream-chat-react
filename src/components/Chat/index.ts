export * from './Chat';
export * from './hooks/useChat';
export * from './hooks/useCustomStyles';
export * from './hooks/useCreateChatClient';
