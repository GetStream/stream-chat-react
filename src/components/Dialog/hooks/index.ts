export * from './useDialog';
export type { PopperLikePlacement } from './usePopoverPosition';
