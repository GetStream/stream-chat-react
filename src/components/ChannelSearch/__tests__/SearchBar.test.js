import React from 'react';
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import { SearchBar } from '../SearchBar';
import { Chat } from '../../Chat';
import { useChannelSearch } from '../hooks/useChannelSearch';

import {
  generateUser,
  getTestClientWithUser,
  queryUsersApi,
  useMockedApis,
} from '../../../mock-builders';

// SearchBar needs searchBarRef to work correctly. That is not expected to be used with theme version 1.
jest.spyOn(window, 'getComputedStyle').mockReturnValue({
  getPropertyValue: jest.fn().mockReturnValue('2'),
});

let client;
const inputText = new Date().getTime().toString();

const AppMenu = ({ close }) => (
  <div>
    AppMenu
    <div data-testid='menu-item' onClick={close} />
  </div>
);
const ClearInputIcon = () => <div>CustomClearInputIcon</div>;
const MenuIcon = () => <div>CustomMenuIcon</div>;
const SearchInputIcon = () => <div>CustomSearchInputIcon</div>;

const SearchContainer = ({ props = {}, searchParams }) => {
  const controller = useChannelSearch(searchParams);
  return <SearchBar {...controller} {...props} />;
};

const renderComponent = ({ client, props = {}, searchParams }) =>
  render(
    <Chat client={client}>
      <SearchContainer props={props} searchParams={searchParams} />
    </Chat>,
  );

describe('SearchBar', () => {
  beforeEach(async () => {
    const user = generateUser();
    client = await getTestClientWithUser({ id: user.id });
    useMockedApis(client, [queryUsersApi([user])]); // eslint-disable-line react-hooks/rules-of-hooks
  });

  it.each([
    ['enable', false, 'xxxxxxxxxx', 'xxxxxxxxxx'],
    ['disable', true, 'xxxxxxxxxx', ''],
  ])('should %s typing', async (_, disabled, inputText, expectedValue) => {
    await renderComponent({ client, searchParams: { disabled } });

    const input = screen.queryByTestId('search-input');

    await act(() => {
      fireEvent.change(input, {
        target: {
          value: inputText,
        },
      });
    });

    await waitFor(() => {
      expect(input).toHaveValue(expectedValue);
    });
  });
  it('should render default layout', () => {
    expect(
      renderer
        .create(
          <SearchBar
            activateSearch={jest.fn}
            clearState={jest.fn}
            exitSearch={jest.fn}
            inputIsFocused={false}
            inputRef={{ current: null }}
            onSearch={jest.fn}
            query=''
            searchBarRef={{ current: null }}
          />,
        )
        .toJSON(),
    ).toMatchSnapshot();
  });
  it.each([
    ['should not render', undefined],
    ['should render', AppMenu],
  ])('%s menu icon', async (_, AppMenu) => {
    await render(
      <SearchBar
        activateSearch={jest.fn}
        AppMenu={AppMenu}
        clearState={jest.fn}
        exitSearch={jest.fn}
        inputIsFocused={false}
        inputRef={{ current: null }}
        onSearch={jest.fn}
        query=''
        searchBarRef={{ current: null }}
      />,
    );
    await waitFor(() => {
      if (!AppMenu) {
        expect(screen.queryByTestId('menu-icon')).not.toBeInTheDocument();
      } else {
        expect(screen.queryByTestId('menu-icon')).toBeInTheDocument();
      }
    });
  });
  it('should render custom icons', async () => {
    await render(
      <SearchBar
        activateSearch={jest.fn}
        AppMenu={AppMenu}
        ClearInputIcon={ClearInputIcon}
        clearState={jest.fn}
        exitSearch={jest.fn}
        inputIsFocused={false}
        inputRef={{ current: null }}
        MenuIcon={MenuIcon}
        onSearch={jest.fn}
        query=''
        searchBarRef={{ current: null }}
        SearchInputIcon={SearchInputIcon}
      />,
    );
    expect(screen.queryByText('CustomClearInputIcon')).toBeInTheDocument();
    expect(screen.queryByText('CustomMenuIcon')).toBeInTheDocument();
    expect(screen.queryByText('CustomSearchInputIcon')).toBeInTheDocument();
  });

  it('should not render ExitSearchIcon if input is not focused', async () => {
    await act(() => {
      renderComponent({ client, searchParams: { disabled: false } });
    });

    await waitFor(() => {
      expect(screen.queryByTestId('return-icon')).not.toBeInTheDocument();
    });
  });

  it('should render ExitSearchIcon on input focus', async () => {
    await renderComponent({ client, searchParams: { disabled: false } });

    const input = screen.queryByTestId('search-input');

    await act(() => {
      fireEvent.focus(input);
    });

    await waitFor(() => {
      expect(screen.queryByTestId('return-icon')).toBeInTheDocument();
    });
  });
  it('should render custom ExitSearchIcon', async () => {
    const ExitSearchIcon = () => <div>CustomExitSearchIcon</div>;
    await renderComponent({
      client,
      props: { ExitSearchIcon },
      searchParams: { disabled: false },
    });

    const input = screen.queryByTestId('search-input');

    await act(() => {
      fireEvent.focus(input);
    });

    await waitFor(() => {
      expect(screen.queryByText('CustomExitSearchIcon')).toBeInTheDocument();
    });
  });
  it('should render custom input placeholder', async () => {
    const placeholder = 'Type and search xxxx';
    await act(() => {
      renderComponent({
        client,
        props: { placeholder },
        searchParams: { disabled: false },
      });
    });

    await waitFor(() => {
      expect(screen.queryByPlaceholderText(placeholder)).toBeInTheDocument();
    });
  });
  it('should clear input', async () => {
    renderComponent({ client, searchParams: { disabled: false } });

    const input = screen.queryByTestId('search-input');

    await act(() => {
      input.focus();
      fireEvent.change(input, {
        target: {
          value: inputText,
        },
      });
    });

    await waitFor(() => {
      expect(input).toHaveValue(inputText);
    });

    const clearButton = screen.queryByTestId('clear-input-button');

    await act(() => {
      fireEvent.click(clearButton);
    });

    await waitFor(() => {
      expect(input).toHaveValue('');
      expect(input).toHaveFocus();
      expect(screen.queryByTestId('return-icon')).toBeInTheDocument();
    });
  });

  it.each([
    [
      'on return button click',
      (target) => {
        fireEvent.click(target);
      },
    ],
    [
      'on Escape key down',
      (target) => {
        fireEvent.keyDown(target, { key: 'Escape' });
      },
    ],
  ])('should exit search UI %s', async (_case, doExitAction) => {
    await renderComponent({ client, searchParams: { disabled: false } });

    const input = screen.queryByTestId('search-input');

    await act(() => {
      input.focus();
      fireEvent.change(input, {
        target: {
          value: inputText,
        },
      });
    });

    await waitFor(() => {
      expect(input).toHaveValue(inputText);
    });

    const returnButton = screen.queryByTestId('search-bar-button');

    await act(() => {
      const target = _case === 'on return button click' ? returnButton : input;
      doExitAction(target);
    });

    await waitFor(() => {
      expect(input).toHaveValue('');
      expect(input).not.toHaveFocus();
      expect(screen.queryByTestId('return-icon')).not.toBeInTheDocument();
    });
  });

  it('should render custom SearchInput', async () => {
    const SearchInput = () => <div>CustomSearchInput</div>;
    await act(() => {
      renderComponent({
        client,
        props: { SearchInput },
        searchParams: { disabled: false },
      });
    });

    await waitFor(() => {
      expect(screen.queryByText('CustomSearchInput')).toBeInTheDocument();
      expect(screen.queryByTestId('search-input')).not.toBeInTheDocument();
    });
  });

  it('should toggle app menu render with menu icon click', async () => {
    await act(() => {
      renderComponent({
        client,
        props: { AppMenu },
        searchParams: { disabled: false },
      });
    });
    const menuIcon = screen.queryByTestId('menu-icon');
    await act(() => {
      fireEvent.click(menuIcon);
    });
    await waitFor(() => {
      expect(screen.queryByText('AppMenu')).toBeInTheDocument();
    });
    await act(() => {
      fireEvent.click(menuIcon);
    });
    await waitFor(() => {
      expect(screen.queryByText('AppMenu')).not.toBeInTheDocument();
    });
  });

  it('should close the app menu on menu item click', async () => {
    await act(() => {
      renderComponent({
        client,
        props: { AppMenu },
        searchParams: { disabled: false },
      });
    });
    const menuIcon = screen.queryByTestId('menu-icon');
    await act(() => {
      fireEvent.click(menuIcon);
    });

    const menuItem = screen.queryByTestId('menu-item');

    await act(() => {
      fireEvent.click(menuItem);
    });

    await waitFor(() => {
      expect(screen.queryByText('AppMenu')).not.toBeInTheDocument();
    });
  });

  it.each([
    [
      'on click outside',
      (target) => {
        fireEvent.click(target);
      },
    ],
    [
      'on Escape key down',
      (target) => {
        fireEvent.keyDown(target, { key: 'Escape' });
      },
    ],
  ])('should close app menu %s', async (_, doCloseAction) => {
    await act(() => {
      renderComponent({
        client,
        props: { AppMenu },
        searchParams: { disabled: false },
      });
    });
    const menuIcon = screen.queryByTestId('menu-icon');
    const searchBar = screen.queryByTestId('search-bar');
    await act(() => {
      fireEvent.click(menuIcon);
    });
    await waitFor(() => {
      expect(screen.queryByText('AppMenu')).toBeInTheDocument();
    });

    await act(() => {
      doCloseAction(searchBar);
    });

    await waitFor(() => {
      expect(screen.queryByText('AppMenu')).not.toBeInTheDocument();
    });
  });
});
