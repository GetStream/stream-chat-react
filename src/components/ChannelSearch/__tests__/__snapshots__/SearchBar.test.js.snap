// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render default layout 1`] = `
<div
  className="str-chat__channel-search-bar"
  data-testid="search-bar"
>
  <div
    className="str-chat__channel-search-input--wrapper"
  >
    <div
      className="str-chat__channel-search-input--icon"
    >
      <svg
        fill="none"
        height="18"
        viewBox="0 0 18 18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.7549 11.255H11.9649L11.6849 10.985C12.6649 9.845 13.2549 8.365 13.2549 6.755C13.2549 3.165 10.3449 0.255005 6.75488 0.255005C3.16488 0.255005 0.254883 3.165 0.254883 6.755C0.254883 10.345 3.16488 13.255 6.75488 13.255C8.36488 13.255 9.84488 12.665 10.9849 11.685L11.2549 11.965V12.755L16.2549 17.745L17.7449 16.255L12.7549 11.255ZM6.75488 11.255C4.26488 11.255 2.25488 9.245 2.25488 6.755C2.25488 4.26501 4.26488 2.255 6.75488 2.255C9.24488 2.255 11.2549 4.26501 11.2549 6.755C11.2549 9.245 9.24488 11.255 6.75488 11.255Z"
          fill="#747881"
        />
      </svg>
    </div>
    <input
      className="str-chat__channel-search-input"
      data-testid="search-input"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <button
      className="str-chat__channel-search-input--clear-button"
      data-testid="clear-input-button"
      disabled={true}
      onClick={[Function]}
    >
      <svg
        fill="none"
        height="14"
        viewBox="0 0 14 14"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14 1.41L12.59 0L7 5.59L1.41 0L0 1.41L5.59 7L0 12.59L1.41 14L7 8.41L12.59 14L14 12.59L8.41 7L14 1.41Z"
          fill="#747881"
        />
      </svg>
    </button>
  </div>
</div>
`;
